cloudfiles/interfaces.py

```diff
@@ -33,13 +33,18 @@ def __missing__(self, key):
 S3_POOL = None
 GC_POOL = None
 MEM_POOL = None
+
+S3ConnectionPoolParams = namedtuple('S3ConnectionPoolParams', 'service bucket_name user_project')
+GCloudBucketPoolParams = namedtuple('GCloudBucketPoolParams', 'bucket_name user_project')
+MemoryPoolParams = namedtuple('MemoryPoolParams', 'bucket_name')
+
 def reset_connection_pools():
   global S3_POOL
   global GC_POOL
   global MEM_POOL
-  S3_POOL = keydefaultdict(lambda service: keydefaultdict(lambda bucket_name: S3ConnectionPool(service, bucket_name)))
```

Comment:
Connections are currently hashed by `bucket_name` (and `service` for S3). If you create a new CloudVolume on the same bucket with a different request payer, the Pool will fetch the existing connection. But I also didn't want to pile on more `keydefaultdicts`. So instead, I am using a namedtuple. Hope, this cleans things up.

Reply:
So if a bucket is requestor pays, then you have to provide your user_project so you are billed correctly? For example, if my billing project is XYZ, then CloudFiles(..., user_project="XYZ"). 

- What happens if user_project is set for non-requestor pays? 
- What if the account doesn't exist or is disabled? 
- What if user_project is required but none is provided? Can you add an error message?
- Would `user_project` be more aptly named `billing_project` or more simply `project`?

Reply:
1) `user_project=None` (default)
a) If the bucket **does not** enforce requester pays, then the bucket host has to pay, same behavior as now
b) If the bucket **does** enforce requester pays. then boto will throw a 403 (Access Denied) and GCS returns 400 with a clear custom error message

2) `user_project="nonsense"`
GCS returns 400 with a clear custom error message, S3 is handled by cloudfiles, because the only allowed values are `None` and `requester`... but I am sure boto would also complain

3) `user_project=None`, while requester pays is strictly enforced on the bucket
That's 1b)

4) No preference here, but `billing_project` would be inaccurate for GCS, they specifically ask for the project name. GCS uses `user_project`, S3/Boto use `RequestPayer`

Reply:
1c) I am throwing an error if the `HTTPInterface` is called with a `user_project` specified - these requests would otherwise go through and the (public read) bucket owner still would be charged, despite the good intentions of the requester.

There is probably a way to set up the HTTPInterface as well by passing authentication tokens, but I haven't looked into that, yet. And we'd probably have to split it into GCSHTTPInterface and S3HTTPInterface(?)

Reply:
So looking more carefully at the s3 and GCS interfaces, it looks like they work on different principles. GCS lets you specify the billing account, but AWS works from the project the secrets token is based on. Can GCS bill to a different account than the token? 

In GCS secret keys, the project is provided. Maybe it would be better to have a top level option `requester_pays=True/False` and then internally AWS is set to "requester" and GCS fetches the user account from the token?

Reply:
GCS seems to be much more flexible there: For example, you could:
First, create a service account on Project A (provides credentials)
Second, add it as IAM to Project B and assign role `Service Usage Consumer` to it (allows to specify project B as request payer)
Third, add it as member to a bucket on Project C with `Storage Object Admin` role. (allows to read/write stuff on project C)

So in that scenario you are writing to project C using credentials from project A, paying from the billing account attached to project C...

Reply:
That's a pretty good feature, but seems excessive for most people. What if we had `requester_pays=True/False` but for GCS you could also instead supply `requester_pays='some-project'`?

Reply:
The reason I am pushing against this inferred `requester_pays` setting is precisely because it is rather powerful (and potentially expensive) feature. I want the user to explicitly write out the project ID so that no unexpected project IDs are used from secret files / environment variables.

Reply:
That's a good point. If that's the case though and we want to be explicit, I think calling the flag `billing` or `requester_pays` is important b/c there's a chance a user will set the flag and it'll keep working even when they change their secret files and conclude it doesn't do anything.

Reply:
Sure, `user_project` is definitely a bit... meaningless.

Reply:
Something is still rubbing me the wrong way about `request_payer` I feel like `billing_id` is more direct. There no operations that CF does that aren't requests so it feels redundant and it's maybe not clear what to put in there. `request_payer="Nico"`? `_id` gives an indication that there's some defined string or number you should look for. I'll leave that up to you though. 

Lastly, would you be able to add some documentation to the README? Thanks again, this is a great PR!